specs/networking/p2p-interface.md

```diff
@@ -0,0 +1,743 @@
+
+### Why are the topics strings and not hashes?
+
+Topics names have a hierarchical structure. In the future, gossipsub may support wildcard subscriptions (e.g. subscribe to all children topics under a root prefix) by way of prefix matching. Enforcing hashes for topic names would preclude us from leveraging such features going forward.
+
+No security or privacy guarantees are lost as a result of choosing plaintext topic names, since the domain is finite anyway, and calculating a digest's preimage would be trivial.
+
+Furthermore, the ETH2 topic names are shorter their digest equivalents (asuming SHA-256 hash), so hashing topics would bloat messages unnecessarily.
+
+### Why are there `SHARD_SUBNET_COUNT` subnets, and why is this not defined?
+
+Depending on the number of validators, it may be more efficient to group shard subnets and might provide better stability for the gossipsub channel. The exact grouping will be dependent on more involved network tests. This constant allows for more flexibility in setting up the network topology for attestation aggregation (as aggregation should happen on each subnet).
+
+### Why are we sending entire objects in the pubsub and not just hashes?
+
+Entire objects should be sent to get the greatest propagation speeds. If only hashes are sent, then block and attestation propagation is dependent on recursive requests from each peer. In a hash-only scenario, peers could receive hashes without knowing who to download the actual contents from. Sending entire objects ensures that they get propagated through the entire network.
+
+### Should clients gossip blocks if they *cannot* validate the proposer signature due to not yet being synced, not knowing the head block, etc?
+
+The prohibition of unverified-block-gossiping extends to nodes that cannot verify a signature due to not being fully synced to ensure that such (amplified) DOS attacks are not possible.
+
+### How are we going to discover peers in a gossipsub topic?
+
+Via discv5 topics. ENRs should not be used for this purpose, as they store identity, location and capability info, not volatile advertisements.
+
+In the interoperability testnet, all peers will be subscribed to all global beacon chain topics, so discovering peers in specific shard topics will be unnecessary.
+
+## Req/Resp
+
+### Why segregate requests into dedicated protocol IDs?
+
+Requests are segregated by protocol ID to:
+
+1. Leverage protocol routing in libp2p, such that the libp2p stack will route the incoming stream to the appropriate handler. This allows each the handler function for each request type to be self-contained. For an analogy, think about how you attach HTTP handlers to a REST API server.
+2. Version requests independently. In a coarser-grained umbrella protocol, the entire protocol would have to be versioned even if just one field in a single message changed.
```

Comment:
counterpoint, how to reliably handle interdependent endpoints (are there any?). If endpoint `A` also needs endpoint `B` upgraded, there is no mechanism to do that reliably with this approach.

Reply:
The plan is to group `A` and `B` in the discovery layer under protocol "families/suites/manifests" -- which represent aggregations of finer-grained protocols under one logical unit. that way we get the best of both worlds: upgraded peers can advertise/negotiate newer versions of a given operation, while continuing to serve older peers by supporting the former versions too.

while we could consider a mechanism to expose machine-readable metadata that specifies that `A` depends on `B`, I believe that's beyond scope. in practice, it should be enough to:

1. specify that in the RFC/EIP that introduces `A`.
2. potentially grouping `A` and `B` under a protocol family that can be atomically advertised.
3. allowing peers to incur in a runtime error when they encounter a peer that supports `A` but not `B`.

Reply:
Cross-referenced from https://github.com/ethereum/eth2.0-specs/issues/1335.